File: workers/common/grpc_metadata.py
"""Utilities for reading SourceBridge gRPC metadata."""

from __future__ import annotations

import json
from dataclasses import dataclass

import grpc


def resolve_model_override(context: grpc.aio.ServicerContext) -> str | None:
    """Read model override from gRPC metadata, return None to use default.

    The Go API attaches 'x-sb-model' metadata when advanced mode is on.
    """
    if not hasattr(context, "invocation_metadata"):
        return None
    for key, value in context.invocation_metadata():
        if key == "x-sb-model":
            return value or None
    return None


@dataclass
class RuntimeLLMOverride:
    provider: str = ""
    base_url: str = ""
    api_key: str = ""
    model: str = ""
    draft_model: str = ""
    operation: str = ""

    def is_empty(self) -> bool:
        return not any(
            (
                self.provider,
                self.base_url,
                self.api_key,
                self.model,
                self.draft_model,
                self.operation,
            )
        )


def resolve_llm_override(context: grpc.aio.ServicerContext) -> RuntimeLLMOverride | None:
    """Read the API-provided effective LLM config from gRPC metadata."""
    if not hasattr(context, "invocation_metadata"):
        return None
    override = RuntimeLLMOverride()
    for key, value in context.invocation_metadata():
        if key == "x-sb-llm-provider":
            override.provider = value or ""
        elif key == "x-sb-llm-base-url":
            override.base_url = value or ""
        elif key == "x-sb-llm-api-key":
            override.api_key = value or ""
        elif key == "x-sb-model":
            override.model = value or ""
        elif key == "x-sb-llm-draft-model":
            override.draft_model = value or ""
        elif key == "x-sb-operation":
            override.operation = value or ""
    if override.is_empty():
        return None
    return override


@dataclass
class RuntimeJobLogMetadata:
    job_id: str = ""
    repo_id: str = ""
    artifact_id: str = ""
    subsystem: str = ""
    job_type: str = ""

    def is_empty(self) -> bool:
        return not any(
            (
                self.job_id,
                self.repo_id,
                self.artifact_id,
                self.subsystem,
                self.job_type,
            )
        )


def resolve_job_log_metadata(context: grpc.aio.ServicerContext) -> RuntimeJobLogMetadata | None:
    """Read job log metadata from gRPC metadata."""
    if not hasattr(context, "invocation_metadata"):
        return None
    meta = RuntimeJobLogMetadata()
    for key, value in context.invocation_metadata():
        if key == "x-sb-job-id":
            meta.job_id = value or ""
        elif key == "x-sb-repo-id":
            meta.repo_id = value or ""
        elif key == "x-sb-artifact-id":
            meta.artifact_id = value or ""
        elif key == "x-sb-subsystem":
            meta.subsystem = value or ""
        elif key == "x-sb-job-type":
            meta.job_type = value or ""
    if meta.is_empty():
        return None
    return meta


@dataclass
class CliffNotesRenderMetadata:
    render_only: bool = False
    selected_section_titles: list[str] | None = None
    understanding_depth: str = ""
    relevance_profile: str = ""

    def is_empty(self) -> bool:
        return (
            not self.render_only
            and not self.selected_section_titles
            and not self.understanding_depth
            and not self.relevance_profile
        )


def resolve_cliff_notes_render_metadata(
    context: grpc.aio.ServicerContext,
) -> CliffNotesRenderMetadata | None:
    if not hasattr(context, "invocation_metadata"):
        return None
    meta = CliffNotesRenderMetadata()
    for key, value in context.invocation_metadata():
        if key == "x-sb-cliff-render-only":
            meta.render_only = (value or "").strip().lower() == "true"
        elif key == "x-sb-cliff-selected-sections":
            raw = (value or "").strip()
            if not raw:
                continue
            try:
                parsed = json.loads(raw)
            except Exception:
                continue
            if isinstance(parsed, list):
                meta.selected_section_titles = [str(item).strip() for item in parsed if str(item).strip()]
        elif key == "x-sb-cliff-understanding-depth":
            meta.understanding_depth = (value or "").strip().lower()
        elif key == "x-sb-cliff-relevance-profile":
            meta.relevance_profile = (value or "").strip().lower()
    if meta.is_empty():
        return None
    return meta

File: internal/api/graphql/grpc_metadata.go
// SPDX-License-Identifier: AGPL-3.0-or-later
// Copyright (C) 2026 SourceBridge Contributors

package graphql

import (
	"context"
	"encoding/json"
	"strconv"
	"strings"

	"google.golang.org/grpc/metadata"

	"github.com/sourcebridge/sourcebridge/internal/llm"
)

// withModelMetadata enriches a context with the API's effective LLM
// configuration for a single worker call. The API owns the DB-backed
// LLM settings, so the worker should not rely on deployment env vars
// as the runtime source of truth.
func (r *Resolver) withModelMetadata(ctx context.Context, operationGroup string) context.Context {
	return r.withJobMetadata(ctx, operationGroup, nil, "", "", "")
}

func (r *Resolver) withJobMetadata(
	ctx context.Context,
	operationGroup string,
	rt llm.Runtime,
	repoID, artifactID, jobType string,
) context.Context {
	if r.Config == nil {
		return ctx
	}

	model := r.Config.LLM.ModelForOperation(operationGroup)
	if operationGroup == "knowledge" && jobType == "architecture_diagram" && r.Config.LLM.ArchitectureDiagramModel != "" {
		model = r.Config.LLM.ArchitectureDiagramModel
	}
	pairs := []string{
		"x-sb-llm-provider", r.Config.LLM.Provider,
		"x-sb-llm-base-url", r.Config.LLM.BaseURL,
		"x-sb-llm-api-key", r.Config.LLM.APIKey,
		"x-sb-llm-draft-model", r.Config.LLM.DraftModel,
		"x-sb-operation", operationGroup,
	}
	if r.Config.LLM.TimeoutSecs > 0 {
		pairs = append(pairs, "x-sb-llm-timeout-seconds", strconv.Itoa(r.Config.LLM.TimeoutSecs))
	}
	if model != "" {
		pairs = append(pairs, "x-sb-model", model)
	}
	if rt != nil && rt.JobID() != "" {
		pairs = append(pairs, "x-sb-job-id", rt.JobID())
	}
	if repoID != "" {
		pairs = append(pairs, "x-sb-repo-id", repoID)
	}
	if artifactID != "" {
		pairs = append(pairs, "x-sb-artifact-id", artifactID)
	}
	if jobType != "" {
		pairs = append(pairs, "x-sb-job-type", jobType)
	}
	pairs = append(pairs, "x-sb-subsystem", "knowledge")
	md := metadata.Pairs(pairs...)
	return metadata.NewOutgoingContext(ctx, md)
}

func withCliffNotesRenderMetadata(
	ctx context.Context,
	renderOnly bool,
	selectedSectionTitles []string,
	understandingDepth string,
	relevanceProfile string,
) context.Context {
	if !renderOnly && len(selectedSectionTitles) == 0 && understandingDepth == "" && relevanceProfile == "" {
		return ctx
	}
	pairs := []string{}
	if renderOnly {
		pairs = append(pairs, "x-sb-cliff-render-only", "true")
	}
	if len(selectedSectionTitles) > 0 {
		if raw, err := json.Marshal(selectedSectionTitles); err == nil {
			pairs = append(pairs, "x-sb-cliff-selected-sections", string(raw))
		}
	}
	if strings.TrimSpace(understandingDepth) != "" {
		pairs = append(pairs, "x-sb-cliff-understanding-depth", strings.TrimSpace(strings.ToLower(understandingDepth)))
	}
	if strings.TrimSpace(relevanceProfile) != "" {
		pairs = append(pairs, "x-sb-cliff-relevance-profile", strings.TrimSpace(strings.ToLower(relevanceProfile)))
	}
	if len(pairs) == 0 {
		return ctx
	}
	md, _ := metadata.FromOutgoingContext(ctx)
	md = md.Copy()
	for i := 0; i < len(pairs); i += 2 {
		md.Set(pairs[i], pairs[i+1])
	}
	return metadata.NewOutgoingContext(ctx, md)
}
